pkg/operator/controller/certificate/default_cert.go

```diff
@@ -17,67 +17,77 @@ import (
 
 // ensureDefaultCertificateForIngress creates or deletes an operator-generated
 // default certificate for a given IngressController as appropriate.  Returns true
-// if it creates, updates, or deletes the secret for the certificate, false
-// otherwise.
+// if it the secret exists, or false if it does not, as well as any errors.
 func (r *reconciler) ensureDefaultCertificateForIngress(caSecret *corev1.Secret, namespace string, deploymentRef metav1.OwnerReference, ci *operatorv1.IngressController) (bool, error) {
 	ca, err := crypto.GetCAFromBytes(caSecret.Data["tls.crt"], caSecret.Data["tls.key"])
 	if err != nil {
 		return false, fmt.Errorf("failed to get CA from secret %s/%s: %v", caSecret.Namespace, caSecret.Name, err)
 	}
-	desired, err := desiredRouterDefaultCertificateSecret(ca, namespace, deploymentRef, ci)
+	wantCert, desired, err := desiredRouterDefaultCertificateSecret(ca, namespace, deploymentRef, ci)
 	if err != nil {
 		return false, err
 	}
-	current, err := r.currentRouterDefaultCertificate(ci, namespace)
+	if !wantCert {
+		// If the operator generated certificate is not being used, ensure that the ingress controller's
+		// Spec.DefaultCertificate secret exists before deleting the operator generated secret.
+		// See https://bugzilla.redhat.com/show_bug.cgi?id=1887441
+		err := r.lookupUserSpecifiedRouterDefaultCertificate(ci, namespace)
+		if err != nil {
+			return false, fmt.Errorf("failed to lookup user specified default certificate: %v", err)
+		}
+	}
+
+	haveCert, current, err := r.currentRouterDefaultCertificate(ci, namespace)
 	if err != nil {
 		return false, err
 	}
 	switch {
-	case desired == nil && current == nil:
+	case !wantCert && !haveCert:
 		// Nothing to do.
-	case desired == nil && current != nil:
+	case !wantCert && haveCert:
 		if deleted, err := r.deleteRouterDefaultCertificate(current); err != nil {
 			return false, fmt.Errorf("failed to delete default certificate: %v", err)
 		} else if deleted {
 			r.recorder.Eventf(ci, "Normal", "DeletedDefaultCertificate", "Deleted default wildcard certificate %q", current.Name)
-			return true, nil
+			return false, nil
```

Comment:
Why is this now considered false?

Reply:
If deletion fails, we technically still have the resource, so we should return true there. If deletion succeeds, however, we should return false since we no long have the resource. Fixed, Thanks!

Reply:
https://github.com/openshift/cluster-ingress-operator/pull/505#discussion_r537829578 explains the inversion here btw.

Reply:
Looks like at the moment the reconciler doesn't even use the boolean!  https://github.com/openshift/cluster-ingress-operator/blob/175d7dfe2f97d0db2196706ea180b2a0e7c4da05/pkg/operator/controller/certificate/controller.go#L105

Reply:
I noticed that as well. But it might in the future!

Reply:
> Maybe it would be a good idea to add a unit test?
Whoops, didn't see this comment before.
We generally don't have unit tests for reconciler methods. Perhaps there should be an e2e test to cover this case?